### PR TITLE
added proper ip checks to l2tunnel, quality of life changes to networ…

### PR DESCRIPTION
**Pull request by jfeiglin**
…k headers. Added the disruption rules interface to the l2tunnel, and a file of disruption rules

### CODE REVIEW

File: disruption_rules.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    return disrupt_ip_traffic
+
+
+def change_dns_responses(domain_name, new_ip):
+    """
+
+    :param domain_name:
+    :param new_ip:
+    :return:
```
Comment (benjaminbaker-dev) [inline disruption_rules.py]: looks like you forgot docstr

File: disruption_rules.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+DNS_PORT = 53
+
+DNS_QUERY = 0
+DNS_RESPONSE = 1
+
+DNS_TYPE_IPV4 = 1
+DNS_TYPE_IPV6 = 28
+
+def redirect_ips(target_ip, redirect_ip):
```
Comment (benjaminbaker-dev) [inline disruption_rules.py]: make name more indicative - i.e ip_redirect_func_factory

File: disruption_rules.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    :param redirect_ip: the source ip to swap to
+    :return: a disruption rule function (takes a header and payload, returns a header and payload)
+    """
+    target_ip = inet_aton(target_ip)
+    redirect_ip = inet_aton(redirect_ip)
+
+    def disrupt_ip_traffic(ip_header, ip_payload):
+        if not isinstance(ip_header, IpHeader):
+            #not an ip packet so ignore
+            return ip_header, ip_payload
```
Comment (benjaminbaker-dev) [inline disruption_rules.py:26]: maybe raise Exception ?

File: disruption_rules.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            ip_header.dst_ip = redirect_ip
+
+        elif ip_header.src_ip == redirect_ip:
+            print('Spoofing packet origin from {} to {}'.format(ip_header.src_ip_str, inet_ntoa(target_ip)))
+            ip_header.src_ip = target_ip
+
+
+
+        #adjust checksums, if they exist
+        if ip_header.proto == IPPROTO_TCP:
```
Comment (benjaminbaker-dev) [inline disruption_rules.py:40]: thought we dont support TCP ?

File: disruption_rules.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        l4_payload = ip_payload[l4_header.length():]
+        l4_header.checksum = 0
+        ip_payload = l4_header.get_raw_header() + l4_payload
+
+        return ip_header, ip_payload
+
+    return disrupt_ip_traffic
+
+
+def change_dns_responses(domain_name, new_ip):
```
Comment (benjaminbaker-dev) [inline disruption_rules.py]: make func name indicative of the fact that it is a factory function

File: disruption_rules.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    :param domain_name:
+    :param new_ip:
+    :return:
+    """
+    domain_name = dnslib.DNSLabel(domain_name)
+    new_ip = dnslib.dns.A(new_ip)
+
+    def disrupt_dns_traffic(udp_header, udp_payload):
+        if not isinstance(udp_header, UdpHeader) or udp_header.src_port != DNS_PORT:
+            return udp_header, udp_payload
```
Comment (benjaminbaker-dev) [inline disruption_rules.py:68]: raise Exception

File: disruption_rules.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+        record = dnslib.DNSRecord.parse(udp_payload)
+        if record.header.qr != DNS_RESPONSE:
+            return udp_header, udp_payload
+        answers = record.rr
```
Comment (benjaminbaker-dev) [inline disruption_rules.py]: what is qr and what is rr

File: tunneler/network_headers.py [hunk 8]
```diff
@@ -11,6 +12,15 @@ class InvalidHWAddr(Exception):
     pass
 
 
+def swap_endianess_16(a):
+    """
+    Swap the endianness of a 16 bit number
+    :param a: number to swap
+    :return: the number with the endianness swapped
+    """
+    return struct.unpack("<H", struct.pack(">H", a))[0]
```
Comment (benjaminbaker-dev) [inline tunneler/network_headers.py:21]: big yeet

File: tunneler/network_headers.py [hunk 9]
```diff
@@ -105,6 +135,9 @@ def dst_addr_str(self):
     def src_addr_str(self):
         return type(self).get_str_from_mac_bytes(self.src_addr)
```
Comment (benjaminbaker-dev) [inline tunneler/network_headers.py:141]: skrt skrt

File: tunneler/tunneler.py [hunk 10]
```diff
@@ -50,6 +85,55 @@ def repackage_frame(self, raw_frame):
...
+        """
+        recv_etherheader = EtherHeader.parse_header(raw_data[:EtherHeader.TOTAL_HEADER_LEN])
+
+        l2_payload = raw_data[recv_etherheader.length():]
+        recv_etherheader, l2_payload = self._disruption_rules[2](recv_etherheader, l2_payload)
+
+        raw_ip_header = l2_payload[:IpHeader.DEFAULT_HEADER_SIZE]
+        recv_ipheader = IpHeader.parse_header(raw_ip_header)
+        l3_payload = l2_payload[recv_ipheader.length():]
+        recv_ipheader, l3_payload = self._disruption_rules[3](recv_ipheader, l3_payload)
```
Comment (benjaminbaker-dev) [inline tunneler/tunneler.py]: make functions for each called _handle_ether_header and _handle_ip_header

File: tunneler/tunneler.py [hunk 11]
```diff
@@ -50,6 +85,55 @@ def repackage_frame(self, raw_frame):
...
+                recv_ipheader.tot_len
+            )
+        elif recv_ipheader.proto == IPPROTO_TCP:
+            recv_l4_header = TcpHeader.parse_header(l3_payload[:TcpHeader.DEFAULT_TCP_HEADER_SIZE])
+            recv_l4_header.pseudo_header = generate_pseudo_header(
+                recv_ipheader.src_ip,
+                recv_ipheader.dst_ip,
+                recv_ipheader.proto,
+                recv_ipheader.tot_len
+            )
```
Comment (benjaminbaker-dev) [inline tunneler/tunneler.py]: take recv_ip_header as parameter to generate_pseudo_header - much better than duplicated unpacking going in now

**Review by benjaminbaker-dev — CHANGES_REQUESTED** [review]
looks super super cool :)

**Comment by benjaminbaker-dev** [comment]
PRIVATE FUNCTIONS

File: disruption_rules.py (hunk 3 above)
Comment (jfeiglin) [reply disruption_rules.py:26]: No you actual idiot can you not read :p... This is called on every packet that we receive, so throwing an exception is essentially dropping the packet and crashing the tunnel

File: disruption_rules.py (hunk 4 above)
Comment (jfeiglin) [reply disruption_rules.py:40]: If you would read the new network headers, you might find that we do

**Review by benjaminbaker-dev — APPROVED** [review]

**Pull request merged**